### PR TITLE
Enables the label content to be an Image.

### PR DESCRIPTION
**Pull request by stockiNail**
This PR is enabling to set the label `content` options as an image.

This could be also helpful when a user would like to apply an text in the label as HTML, see #318.

It's also adding new options, `resizeRatio` (maybe the name is not the best) which can allow to the user to resize the image in the label, by a ratio (percentage).

### CODE REVIEW

**Comment by kurkle** [comment]
I'd think the Github mark is copyrighted, not sure if we should use that.

Not sure about the `resizeRatio`. Size in pixels could be more usable generally.
What do you think about having `width` and `height` that can be pixels or percent?

**Comment by stockiNail** [comment]
> I'd think the Github mark is copyrighted, not sure if we should use that.

I think it could be used if I had added the link to Github project, but to avoid any doubt, it has been removed.

**Comment by stockiNail** [comment]
I've tried to add the test case how you explained me but I can not understand why it doesn't work.

```
[test-karma] Chrome 88.0.4324.96 (Linux x86_64) Line annotation auto /base/test/fixtures/line/contentImage.js FAILED
[test-karma] 	Error: [object HTMLDivElement]
[test-karma] 	    at <Jasmine>
[test-karma] 	    at test/node_modules/chartjs-test-utils/dist/chartjs-test-utils.js:955:23 <- test/index.js:770:25
[test-karma] 	    at Image.image.onload (test/node_modules/chartjs-test-utils/dist/chartjs-test-utils.js:527:5 <- test/index.js:364:7)
[test-karma] Chrome 88.0.4324.96 (Linux x86_64): Executed 6 of 9 (1 FAILED) (0 secs / 0.541 secs)
```

**Comment by kurkle** [comment]
> I've tried to add the test case how you explained me but I can not understand why it doesn't work.
> 
> ```
> [test-karma] Chrome 88.0.4324.96 (Linux x86_64) Line annotation auto /base/test/fixtures/line/contentImage.js FAILED
> [test-karma] 	Error: [object HTMLDivElement]
> [test-karma] 	    at <Jasmine>
> [test-karma] 	    at test/node_modules/chartjs-test-utils/dist/chartjs-test-utils.js:955:23 <- test/index.js:770:25
> [test-karma] 	    at Image.image.onload (test/node_modules/chartjs-test-utils/dist/chartjs-test-utils.js:527:5 <- test/index.js:364:7)
> [test-karma] Chrome 88.0.4324.96 (Linux x86_64): Executed 6 of 9 (1 FAILED) (0 secs / 0.541 secs)
> ```

run the 'dev' + debug again to see a detailed error

**Comment by kurkle** [comment]
> If ok, I'd like to use a regex to get the value when set a percentage. Any constraint to use regex?

A constant regex would be fine. I think it could be something simple too:
```js
> const toPercent = (s) => typeof s === 'string' && s.endsWith('%') && parseFloat(s) / 100;
> toPercent('50.1 %')
> 0.501
```

**Comment by stockiNail** [comment]
> run the 'dev' + debug again to see a detailed error

Strange.. no error with run debug

Capture

**Comment by kurkle** [comment]
What kind of test? Maybe the image is not loaded before trying to draw it?

**Comment by stockiNail** [comment]
> A constant regex would be fine. I think it could be something simple too:

I don't know if you see the code.

I have created the regex (constant for all line.js) and a function to calculate image size:

```javascript
const percentageRegex = RegExp('(\\d+(\\.\\d+)?|\\.\\d+)?%');
.....
function getImageSize(size, value) {
  if (typeof value === 'number') {
    return value;
  } else if (typeof value === 'string') {
    const array = percentageRegex.exec(value);
    if (array) {
      return +array[1] / 100 * size;
    }
  }
  return size;
}
...

Let me know if you prefer to have something string base instead of regex.

**Comment by stockiNail** [comment]
> 
> 
> What kind of test? Maybe the image is not loaded before trying to draw it?

**/test/fixtures/line/contentImage.js**

I have the image creation before module export:

```javascript
const htmlText = new Image();
htmlText.src = 'data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" width="150" height="80">' +
'<foreignObject width="100%" height="100%">' +
'<div xmlns="http://www.w3.org/1999/xhtml" style="color: white;">Here is a <strong style="color: black;">HTML</strong> text, used as image, which can be styled as <em style="font-size: 16px;">YOU</em> like!</div>' +
'</foreignObject></svg>';
module.exports = {
```

This is the only difference comparing with the others (I did the test case for the other PR and it works).

**Comment by stockiNail** [comment]
> 
> 
> > If ok, I'd like to use a regex to get the value when set a percentage. Any constraint to use regex?
> 
> A constant regex would be fine. I think it could be something simple too:
> 
> ```js
> > const toPercent = (s) => typeof s === 'string' && s.endsWith('%') && parseFloat(s) / 100;
> > toPercent('50.1 %')
> > 0.501
> ```

I'm testing it, removing REGEX going to string toPercent.

**Comment by kurkle** [comment]
I think you'd still need to wait for the loading of the image.
add `await img.decode();` after setting the `src`

ref: https://stackoverflow.com/a/64747517/10359775

**Comment by stockiNail** [comment]
> add `await img.decode();` after setting the `src`

I'm trying but `await` needs async function.

File: src/types/line.js [hunk 1]
```diff
@@ -198,15 +198,7 @@ function drawLabel(ctx, line, chartArea) {
...
-  if (typeof value === 'number') {
-    return value;
-  } else if (typeof value === 'string') {
-    const array = percentageRegex.exec(value);
-    if (array) {
-      return +array[1] / 100 * size;
-    }
-  }
-  return size;
+  return typeof value === 'number' ? value : typeof value === 'string' ? toPercent(value) * size : size;
```
Comment (kurkle) [inline src/types/line.js:201]: I think your original function body was better :)

```js
if (typeof value === 'number') {
  return value;
}
if (typeof value === 'string' && value.endsWith('%')) {
  return parseFloat(value) / 100 * size;
}
return size;
```

or (less readable, but slightly smaller footprint due to single `return`)

```js
function getImageSize(size, value) {
	return typeof value === 'number' ? value
		: (typeof value === 'string' && value.endsWith('%')) ? parseFloat(value) / 100 * size
			: size;
}
```

Comment (stockiNail) [reply src/types/line.js:201]: I have changed for that reason, to reduce the size but I can change it quickly

Comment (stockiNail) [reply src/types/line.js]: done

**Review by kurkle — APPROVED** [review]

**Pull request merged**